extension/js/api.js

```diff
@@ -0,0 +1,95 @@
+import $ from 'jquery';
+import endpoints from './api_endpoints.json';
+
+var api_url = (Jupyter.notebook_list || Jupyter.notebook).base_url + "api/packagemanager/";
+
+/*
+This function makes the API calls to the specified endpoint with the corresponding headers
+*/
+
+function api_call(endpoint, url, payload, success) {
+
+    var settings = {
+        url: api_url + endpoint.uri + url,
+        method: endpoint.method,
+        headers: {
+            "Content-Type": "application/json",
+            "cache-control": "no-cache",
+        },
+        processData: false,
+        success: success
+    };
+
+    // Send the payload only if it is not empty
+
+    if (!(Object.keys(payload).length === 0 && payload.constructor === Object)) {
+        settings.data = JSON.stringify(payload);
```

Comment:
This is looking good. But I have to ask *again*: why do you stringify?? you could use the json object, right?

Reply:
I did try that. Apparently, removing that breaks the application. Any hints on what I'll have to investigate?

Reply:
From the doc:

> data
Type: PlainObject or String or Array
Data to be sent to the server. It is converted to a query string, if not already a string. It's appended to the url for GET-requests. See processData option to prevent this automatic processing. Object must be Key/Value pairs. If value is an Array, jQuery serializes multiple values with same key based on the value of the traditional setting (described below).

So I don't know what might be the problem. Do you always pass at least an empty object?

Reply:
Can you try something like:

```
var settings = {
        url: api_url + endpoint.uri + url,
        method: endpoint.method,
        headers: {
            "Content-Type": "application/json",
            "cache-control": "no-cache",
        },
        data: payload || {},
        processData: false,
        success: success,
    };
```